pkg/typesystem/connected_types.go

```diff
@@ -0,0 +1,142 @@
+package typesystem
+
+import (
+	openfgav1 "github.com/openfga/api/proto/openfga/v1"
+)
+
+// map[objectType]map[relation]map[subjectType]terminalRelation.
+type TypesystemConnectedTypes map[string]map[string]map[string][]string
```

Comment:
I don't love this data structure to this information 😓 which is why I used a graph data structure here: https://github.com/openfga/openfga/pull/1582

A graph ds enables us to extend the functionality to find cycles and traverse the model in any way we want

Reply:
I'm open to the idea of modifying the data structure. However, the objective of this work is erformant lookups for terminal types and terminal relations for a given userset. The graph doesn't appear to be any more performant than traversing the typesystem itself, in which case we wouldn't store this data as a cache to begin with. The PR you link eludes to this which is why memoizing is recommended as a means of preventing duplicate function calls, which suggests storing the results in memory ultimately. That's where this PR immediately starts, without the intermediate steps.

Reply:
Agreed offline to keep this as is